wurst/systems/units/buildings/TeleportBeacon.wurst

```diff
@@ -0,0 +1,108 @@
+package TeleportBeacon
+import Building
+import LinkedList
+import HashMap
+import ClosureEvents
+import UnitExtensions
+import TerrainUtils
+import SimError
+
+
+constant let SPELL_TELEPORT_BEACON = 'A042'
+constant let rangeCheckModFirst = "Doodads\\Cinematic\\GlowingRunes\\GlowingRunes2.mdl"
+constant let mirrorEfx = "Abilities\\Spells\\NightElf\\Blink\\BlinkTarget.mdl"
+
+
+public class TeleportBeacon extends Building
+    private static let unitTypeId = UNIT_TELEPORTATION_BEACON
+
+    static function initialize()
+        register(unitTypeId, building -> new TeleportBeacon(building))
+
+    private let effects = new IterableMap<player, LinkedList<effect>>()
+
+    construct(unit whichUnit)
+        super(whichUnit)
+        EventListener.add(whichUnit, EVENT_UNIT_SELECTED, -> onSelected())
+        EventListener.add(whichUnit, EVENT_UNIT_DESELECTED, -> onDeselected())
+        EventListener.add(whichUnit, EVENT_UNIT_PICKUP_ITEM, -> onPickupItem())
+        EventListener.add(whichUnit, EVENT_UNIT_SPELL_CAST, -> onCast())
+
+    override function getTypeId() returns int
+        return unitTypeId
+
+    private function onPickupItem()
+        let pickedUp = GetManipulatedItem()
+        if pickedUp.getTypeId() != ITEM_MAGIC
+            return
+
+        let beacon = getUnit()
+        if beacon.getAbilityLevel(SPELL_TELEPORT_BEACON) > 1
+            return
+
+        pickedUp.remove()
+        beacon.incAbilityLevel(SPELL_TELEPORT_BEACON)
+
+    private function onSelected()
+        let whichPlayer = GetTriggerPlayer()
+
+        if effects.has(whichPlayer)
+            return
+
+        let beacon = getUnit()
+        if not beacon.getOwner().isAllyOf(whichPlayer)
+            return
+
+        let range = beacon.getAbilityLevel(SPELL_TELEPORT_BEACON) == 1 ? 700.0 : 1100.0
```

Comment:
The various ranges could be a constant array I think. That or two constants. Whatever is best practices for configured spells.